### PR TITLE
[2/11] Correct file and RPC semantics

### PR DESCRIPTION
**Pull request by ArthurHeymans**
## Summary

Fix offset writes, directory counts, executable checks, error mapping, dangling symlinks, dynamic groups, and bounded reads.

## Stack

Part **2 of 11** in the TRAMP RPC remediation stack.

- Base: `remediation/01-validation`
- Next PRs build on `remediation/02-file-rpc-semantics`

Each PR contains only the incremental changes since its base branch.

## Full-stack validation

- Rust: 105/105 tests
- Mock ERT: 229/229
- Protocol ERT: 8/8
- Server ERT: 16/16
- Autoload ERT: 11/11
- Remote integration: 99 passed, 3 expected skips, 0 unexpected
- Upstream TRAMP: 73 passed, 38 feature-dependent skips, 0 unexpected
- Rustfmt, Clippy, strict byte compilation, and static musl build passed

Independent Terra validation and Luna adversarial review completed with no remaining blockers on the complete stack.

## Summary by CodeRabbit

* **New Features**
  * Added bounded, chunked remote file reads (range-aware/pipelined) with server-exposed `max_read_chunk_bytes`.
  * Added a new TRAMP-RPC protocol error code for process-related failures.

* **Bug Fixes**
  * Stricter validation for remote `file.read`, `process.read`, and PTY reads; improved chunked fallback on oversized-range rejection.
  * Improved RPC error robustness with structured `code`/`data`, better timeout/error signaling, and hardened spawn/result handling.
  * Offset `file.write` flushes reliably, supports creating missing targets, and preserves suffix bytes; improved rename no-overwrite and grouped COUNT semantics.

* **Tests**
  * Extended coverage for chunking, trash growth retry, offset write/rename edge cases, groups, COUNT handling, and process spawn/pipe/timeout behaviors.

### CODE REVIEW

**Comment by ArthurHeymans** [comment]
Adjusted the read limits in the introducing commit: file reads now use 16 MiB as a per-RPC chunk size and transparently continue across chunks; process and PTY reads have a separate 1 MiB maximum with the existing 64 KiB default. Large optimized trash copies also use the chunked path. Added mock and real RPC boundary coverage. Validation: Rust 105/105, mock ERT 229/229, protocol 8/8, server 16/16, strict byte compilation, Clippy, rustfmt, static musl build, focused remote chunk test, and a manual 16 MiB + 1 byte remote read.

**Comment by ArthurHeymans** [comment]
Adversarial review remediation pushed:
- No-overwrite rename now uses atomic `renameat2(RENAME_NOREPLACE)` on Linux with `EEXIST` mapped to already-exists (check-then-rename kept only as non-Linux fallback), closing the TOCTOU race.
- `executable_is_missing` preflight resolves relative `PATH` entries against the requested `cwd` instead of the server cwd; regression added.

Validation: cargo tests green (74 at this commit, 113 at stack tip).

File: server/src/handlers/io.rs [hunk 1]
```diff
@@ -451,6 +487,40 @@ async fn remove_path_for_overwrite(path: &Path) -> std::io::Result<()> {
...
+#[cfg(not(target_os = "linux"))]
+async fn rename_no_overwrite(src: &Path, dest: &Path) -> std::io::Result<()> {
+    match fs::symlink_metadata(dest).await {
+        Ok(_) => return Err(already_exists(dest)),
+        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {}
+        Err(e) => return Err(e),
+    }
+    fs::rename(src, dest).await
+}
+
```
[the remark this answers is not in the document]
Comment (ArthurHeymans) [reply server/src/handlers/io.rs:562]: Fixed in b1e7cab2. The Linux renameat2 syscall now runs in tokio::task::spawn_blocking. Executable/cwd/PATH metadata checks also run in spawn_blocking, and only after spawn fails, so successful process starts no longer do the preflight at all. The non-Linux rename fallback already uses tokio::fs, which offloads filesystem work internally.

File: server/src/handlers/io.rs [hunk 2]
```diff
@@ -471,18 +541,18 @@ pub async fn rename(params: Value) -> HandlerResult {
...
+    } else {
+        rename_no_overwrite(&src, &dest).await
+    };
+    result.map_err(|e| {
+        if e.kind() == std::io::ErrorKind::AlreadyExists {
+            map_io_error(already_exists(&dest), &dest_str)
+        } else {
+            map_io_error(e, &src_str)
+        }
+    })?;
```
[the remark this answers is not in the document]
Comment (ArthurHeymans) [reply server/src/handlers/io.rs:597]: Fixed in b1e7cab2. Non-AlreadyExists rename errors now include both source and destination (`src -> dest`) while preserving the original I/O error kind/errno.

File: server/src/handlers/mod.rs [hunk 3]
```diff
@@ -165,19 +165,56 @@ fn system_statvfs(params: Value) -> HandlerResult {
...
+    F: FnMut(&mut [libc::gid_t]) -> std::io::Result<usize>,
+{
+    let mut count = initial_count;
+    loop {
+        let mut groups = vec![0; count];
+        match getgroups(&mut groups) {
+            Ok(actual_count) => {
+                groups.truncate(actual_count);
+                return Ok(groups);
+            }
```
[the remark this answers is not in the document]
Comment (ArthurHeymans) [reply server/src/handlers/mod.rs:187]: Fixed in b1e7cab2. An actual count larger than the current buffer now resizes directly to that count and retries. Added a zero-length sizing-call regression that reproduces the race.

**Comment by ArthurHeymans** [comment]
Addressed the remaining CodeRabbit review items in `b1e7cab2`:

- Offloaded Linux `renameat2` and executable lookup filesystem checks from Tokio workers. Executable classification now runs only on spawn failure; the non-Linux rename path already uses `tokio::fs`.
- Rename failures other than `AlreadyExists` report both source and destination.
- Supplementary-group sizing retries when `getgroups` returns a count larger than the supplied buffer, including the zero-length sizing-call race; regression added.
- Consolidated the two duplicate unit tests in the mock suite. Later pending-request assertions were moved to the retained mock test while resolving the stack.

Validation at PR commit: Rust 75/75, protocol 8/8, mock 181/181, strict byte-compile, rustfmt. Final rebased stack tip: Rust 114/114, protocol 8/8, mock 233/233, strict byte-compile, no conflicts.

Skipped the generic 80% docstring-coverage warning: it would require adding metric-only prose to private helpers/tests and is not a project validation gate.
